core/Pos/Binary/Crypto.hs

```diff
@@ -43,263 +34,192 @@ import           Pos.Crypto.Signing         (ProxyCert (..), ProxySecretKey (..)
-constantSizedBinaryToStoreGet :: Binary.Binary a => Int -> Store.Peek a
-constantSizedBinaryToStoreGet bytes = do
-    x <- getBytes bytes
-    case Binary.decodeOrFail (BSL.fromStrict x) of
-        Left (_, _, err) -> fail err
-        Right (bs, _, res)
-            | BSL.null bs -> pure res
-            | otherwise   -> fail "unconsumed input"
-
--- [CSL-1122] TODO: more efficient 'put' and 'get' are possible if 'Store'
--- instances are added into pvss-haskell
-#define BiPvss(T, PT, BYTES) \
+#define BiPvss(T, PT) \
   instance Bi T where {\
-    size = ConstSize BYTES ;\
-    put = labelP "T" . putBytes . BSL.toStrict . Binary.encode ;\
-    get = label "T" $ constantSizedBinaryToStoreGet BYTES };\
+    encode = encodeBinary ;\
+    decode = decodeBinary };\
   deriving instance Bi PT ;\
 
-BiPvss (Pvss.PublicKey, VssPublicKey, 33)    -- yes it's 33 and not 32
-BiPvss (Pvss.KeyPair, VssKeyPair, 65)        -- 32+33
-BiPvss (Pvss.Secret, Secret, 33)
-BiPvss (Pvss.DecryptedShare, Share, 101)     -- 4+33+64
-BiPvss (Pvss.EncryptedShare, EncShare, 101)
-BiPvss (Pvss.Proof, SecretProof, 64)
-
-instance Store.Store Pvss.ExtraGen where
-    size = ConstSize 33
-    poke = labelP "Pvss.ExtraGen" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.ExtraGen" $ constantSizedBinaryToStoreGet 33
-instance Store.Store Pvss.Commitment where
-    size = ConstSize 33
-    poke = labelP "Pvss.Commitment" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.Commitment" $ constantSizedBinaryToStoreGet 33
-
-Store.makeStore ''SecretSharingExtra
+BiPvss (Pvss.PublicKey, VssPublicKey)
+BiPvss (Pvss.KeyPair, VssKeyPair)
+BiPvss (Pvss.Secret, Secret)
+BiPvss (Pvss.DecryptedShare, Share)
+BiPvss (Pvss.EncryptedShare, EncShare)
+BiPvss (Pvss.Proof, SecretProof)
+
+instance Bi Pvss.ExtraGen where
+    encode = encodeBinary
+    decode = decodeBinary
+
+instance Bi Pvss.Commitment where
+    encode = encodeBinary
+    decode = decodeBinary
+
 instance Bi SecretSharingExtra where
-    put = labelP "SecretSharingExtra" . Store.poke
-    get = label "SecretSharingExtra" $ Store.peek
-    size = Store.size
+    encode (SecretSharingExtra eg comms) = encodeListLen 2
+                                        <> encode eg
+                                        <> encode comms
+    decode = SecretSharingExtra
+          <$  enforceSize "SecretSharingExtra" 2
+          <*> decode
+          <*> decode
 
 deriving instance Bi (AsBinary SecretSharingExtra)
 
 ----------------------------------------------------------------------------
 -- SecretSharing AsBinary
 ----------------------------------------------------------------------------
 
-#define BiMacro(B, BYTES) \
+#define BiMacro(B) \
   instance Bi (AsBinary B) where {\
-    size = ConstSize BYTES ;\
-    put (AsBinary bs) = putBytes bs ;\
-    get = label "B (BYTES bytes)" $ AsBinary <$> getBytes BYTES}; \
+    encode (AsBinary bs) = encode bs ;\
+    decode = AsBinary <$> decode}; \
```

Comment:
There are two things which I am concerned with regarding `AsBinary`: unnecessary overhead and handling of malformed data. I've already expressed my thoughts about unnecessary overhead below. Handling of malformed data is tricky. I reviewed the code and I see that for types like `VssPublicKey`, `Secret`, etc. (for which we know exact size and the size is the only check) we ensure that `fromBinary` can never fail by checking the length in `Bi.get`. However, this check is lost in this PR. So now it is possible that `AsBinary VssPublicKey` contains `ByteString` with length ≠ 33. It means that subsequent `fromBinary` call can fail. I am not sure whether the current code will handle this change.
There is also `SecretSharingExtra` for which we the size is variable (it contains a list inside). For this type we don't ensure that `fromBinary` can't fail. I am not sure if it's a bug or not.
I am going to review how higher-level code handles `fromBinary` failures and then I will say whether we need size check here.